core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/AbstractFederationConnection.java

```diff
@@ -434,10 +439,27 @@ private TupleExpr optimize(TupleExpr parsed, Dataset dataset, BindingSet binding
 		new QueryMultiJoinOptimizer().optimize(query, dataset, bindings);
 		// new FilterOptimizer().optimize(query, dataset, bindings);
 
-		new EmptyPatternOptimizer(members).optimize(query, dataset, bindings);
+		// prepare bloom filters
+		RepositoryBloomFilter defaultBloomFilter = new AccurateRepositoryBloomFilter(includeInferred);
+		Map<Repository, RepositoryBloomFilter> bloomFilters = federation.getBloomFilters();
+		Map<RepositoryConnection, RepositoryBloomFilter> bloomFiltersByConn = new HashMap<>();
+		List<Repository> repoMembers = federation.getMembers();
+		for(int i=0; i<members.size(); i++) {
```

Comment:
I agree, but supporting the existing constructors makes it a bit awkward. It is not just the optimizers, but also the connection subclasses. This was less invasive. Ideally, I would want to add a new constructor which took a list of pair<Repository,RepositoryConnection> instead of list<RepoConn>. I can map repoconn -> pair<null,repoconn> to keep backwards compatibility (and I have its adjoint, the forgetful functor in the opposite direction). Every repo is then known (or not) for every repoconn so it is a straightforward lookup to get the filter (or anything else in the future).